tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -534,4 +499,74 @@ void initTorchFunctions(PyObject* module) {
   }
 }
 
+PyObject* handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyTypeObject &torch_api) {
+  PyObject* torch_api_function =
+    PyObject_FastGetAttrString((PyObject*)&torch_api, const_cast<char*>(r.get_func_name().data()));
+    PyObject* ret;
+    // there must be at least one overloaded argument at this point, since r.has_torch_function() is true
+    // so ret will never be returned unset
+    for (auto it = r.overloaded_args.begin(); it != r.overloaded_args.end(); ++it) {
+      PyObject* torch_function = PyObject_FastGetAttrString(*it, "__torch_function__");
+      ret = PyObject_CallFunctionObjArgs(torch_function, torch_api_function, args, kwargs, NULL);
+      if (ret != Py_NotImplemented) {
+        return ret;
+      }
```

Comment:
Leaking reference to `Py_NotImplemented` here

Reply:
also here, can you explain in a little more detail how I'd do this properly?

Reply:
Let me first talk about how you would do it correctly "by hand", and then tell you how pybind11 helps.

Python C API is manually refcounted. This means that you are responsible for manually incref'ing/decref'ing when you acquire/release ownership on a Python object. The mental checklist I use whenever make a C API call looks something like this:

1. (Rare) Does the C API function steal a reference? If so, ensure that I have moved out ownership. If I continue to use the object afterwards, I must incref. Example API function like this: https://docs.python.org/2/c-api/tuple.html#c.PyTuple_SetItem
2. Does the C API function return a new reference? If so, I must ensure that I dispose of the ownership somehow later, e.g., by decref'ing, or passing ownership to someone else (e.g., by returning it or giving it to a function that steals references). Example: https://docs.python.org/2/c-api/tuple.html#c.PyTuple_New

A common mistake is to forget to refcount singleton objects like `Py_None`. All objects in Python are refcounted and you must maintain accurate counts even for objects that stay live for the entirety of the Python program.

pybind11 helps because it provides RAII wrapper that will automatically decref an object when it goes out of scope. https://github.com/pybind/pybind11/issues/1201